src/storage/routing/RegexRouterRule.ts

```diff
@@ -0,0 +1,56 @@
+import type { Representation } from '../../ldp/representation/Representation';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import { UnsupportedHttpError } from '../../util/errors/UnsupportedHttpError';
+import { trimTrailingSlashes } from '../../util/Util';
+import type { ResourceStore } from '../ResourceStore';
+import { RouterRule } from './RouterRule';
+
+interface RegexEntry {
+  regex: RegExp;
+  store: ResourceStore;
+}
+
+/**
+ * Routes requests to a store based on the path of the identifier.
+ * The identifier will be stripped of the base URI after which regexes will be used to find the correct store.
+ * The trailing slash of the base URI will still be present so the first character a regex can match would be that `/`.
+ * This way regexes such as `/container/` can match containers in any position.
+ *
+ * In case none of the regexes match an error will be thrown.
+ */
+export class RegexRouterRule extends RouterRule {
+  private readonly base: string;
+  private readonly regexes: RegexEntry[];
+
+  /**
+   * The keys of the `storeMap` will be converted into actual RegExp objects that will be used for testing.
+   */
+  public constructor(base: string, storeMap: { [regex: string]: ResourceStore }) {
+    super();
+    this.base = trimTrailingSlashes(base);
+    this.regexes = Object.keys(storeMap).map((regex): RegexEntry =>
+      ({ regex: new RegExp(regex, 'u'), store: storeMap[regex] }));
+  }
+
+  public async canHandle(input: { identifier: ResourceIdentifier; representation?: Representation }): Promise<void> {
+    const path = this.toRelative(input.identifier);
+    if (!this.regexes.some((entry): boolean => entry.regex.test(path))) {
+      throw new UnsupportedHttpError(`No stored regexes match ${input.identifier.path}`);
+    }
```

Comment:
Just as a thought: this is a case where `canHandle`/`handle` cam be expensive indeed, because we're finding the regex twice.

However, there can (and actually, probably should, to avoid the current double code) be a member method `findRegex` which returns the right one. 

And then that method, if needed, can be memoized. (I wouldn't do that in this PR yet, but I would do the member method, especially given the double calling of `toRelative`).